### PR TITLE
feat: Test Kit for Servers

### PR DESCRIPTION
**Pull request by yhl25**
* Test kits for servers to simplify unit testing and component testing
* Capability to run the server locally 

## Unit Testings
Just test functions without starting neither server or the client

## Component Testing - Client code
In this mode, server will be already running.
1. start client
2. using client call the server
3. close the client
```
       // Create a client that can send requests to the server
        MapperTestKit.Client client = new MapperTestKit.Client();
        MapperTestKit.TestDatum datum = MapperTestKit.TestDatum
                .builder()
                .value("2".getBytes())
                .build();
        MessageList result = client.sendRequest(new String[]{}, datum);
        client.close();
```

## Server code for just testing the UDFs locally
1. start server
2. open client
3. using client invoke the RPCs functions on server
4. close client
5. showdown server
```
        // Start the server locally. This will initiate the server on localhost and port 50051
        Server server = new Server(new EvenOddFunction());
        // Start the server
        server.start();
        // Wait for the server to shutdown
        server.awaitTermination();
```

### CODE REVIEW

**Review by vigith — APPROVED** [review]

File: examples/pom.xml [hunk 1]
```diff
@@ -110,7 +123,7 @@
                                 </mainClass>
                             </container>
                             <to>
-                                <image>numaflow-java-examples/even-odd:${docker.tag}</image>
+                                <image>quay.io/yhl25/numaflow-java/even-odd:v0.0.1</image>
```
Comment (nagarajatantry) [inline examples/pom.xml:126]: revert?

**Review by KeranYang — COMMENTED** [review]
Thanks @yhl25  for making the changes, overall it LGTM. I made some nit changes.

Not sure if it's just me, the ServerTest under examples folder is broken. When I ran `mvn clean install` under the examples folder, got the following error:

```
[ERROR] Failures: 
[ERROR]   ServerTest.testMapServerInvocation:38 Failed to start server
[ERROR]   ServerTest.testReduceServerInvocation:104 Failed to start server
[ERROR]   ServerTest.testSinkServerInvocation:160 Failed to start server
[ERROR]   ServerTest.testSourceServerInvocation:204 Failed to start server
[ERROR] Errors: 
[ERROR]   ServerTest.testFlatMapServerInvocation:77 » Runtime java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: UNKNOWN
[ERROR]   EventTimeFilterFunctionTest.testServerInvocation:30 » Runtime java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: UNKNOWN
```

Please check...

**Review by nagarajatantry — APPROVED** [review]

File: examples/pom.xml [hunk 2]
```diff
@@ -25,16 +25,27 @@
             <artifactId>slf4j-simple</artifactId>
             <version>2.0.7</version>
         </dependency>
+
```
Comment (KeranYang) [inline examples/pom.xml:28]: `mvn clean install` under examples folder seems broken. could you please fix it before merging?

**Review by KeranYang — CHANGES_REQUESTED** [review]

File: examples/pom.xml (hunk 1 above)
Comment (KeranYang) [reply examples/pom.xml]: done.

File: examples/pom.xml (hunk 2 above)
Comment (nagarajatantry) [reply examples/pom.xml:28]: same here. getting the below error 
```
ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/reducesession/counter/CountFactory.java:[22,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.sessionreducer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/sideinput/simple/SimpleSideInput.java:[37,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.sideinput.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/reduce/sum/SumFactory.java:[18,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.reducer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/reduce/count/CounterFactory.java:[32,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.reducer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/sourcetransformer/eventtimefilter/EventTimeFilterFunction.java:[32,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.sourcetransformer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/mapstream/flatmapstream/FlatMapStreamFunction.java:[26,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.mapstreamer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/source/simple/SimpleSource.java:[37,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.sourcer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/sink/simple/SimpleSink.java:[26,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.sinker.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/map/forward/ForwardFunction.java:[21,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.mapper.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/reducestreamer/sum/SumFactory.java:[22,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.reducestreamer.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/sideinput/udf/SimpleMapWithSideInput.java:[47,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.mapper.Server
[ERROR] /Users/ntantry/dev/intuit_projects/numaflow-java/examples/src/main/java/io/numaproj/numaflow/examples/map/flatmap/FlatMapFunction.java:[25,15] cannot find symbol
  symbol:   method awaitTermination()
  location: variable server of type io.numaproj.numaflow.mapper.Server

```

Comment (nagarajatantry) [reply examples/pom.xml:28]: pls ignore my above comment. Running `mvn clean install` in the parent folder, got rid of this error message. Now I see the same error message as @KeranYang

File: examples/pom.xml [hunk 3]
```diff
@@ -25,16 +25,27 @@
...
         <dependency>
             <groupId>org.mockito</groupId>
             <artifactId>mockito-core</artifactId>
             <version>4.8.1</version>
             <scope>test</scope>
         </dependency>
+
+        <!-- https://mvnrepository.com/artifact/junit/junit -->
+        <dependency>
```
Comment (nagarajatantry) [inline examples/pom.xml:37]: dont need this if running junit 5 (jupiter)

File: examples/src/test/java/io/numaproj/numaflow/examples/server/ServerTest.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+    @Test
+    public void testMapServerInvocation() {
+        MapperTestKit mapperTestKit = new MapperTestKit(new EvenOddFunction());
+        try {
+            mapperTestKit.startServer();
+        } catch (Exception e) {
+            log.error("Failed to start server", e);
+            Assertions.fail("Failed to start server");
+        }
```
Comment (nagarajatantry) [inline examples/src/test/java/io/numaproj/numaflow/examples/server/ServerTest.java:39]: safer to implemente `try-with-resources`. can be tackled in the next PR

File: examples/pom.xml (hunk 3 above)
Comment (yhl25) [reply examples/pom.xml]: done

File: examples/pom.xml (hunk 2 above)
Comment (yhl25) [reply examples/pom.xml:28]: I have fixed it. Please let me know if you still see the error.

Comment (nagarajatantry) [reply examples/pom.xml:28]: working fine now.

**Review by nagarajatantry — APPROVED** [review]
lgtm

**Pull request merged**